File: packages/shared/utils/reel-composition.ts
/**
 * Règle de composition d'un RÉEL — prédicat PUR, SOURCE UNIQUE partagée.
 *
 * Règle produit (directive user 2026-08-02) : un post n'est un RÉEL que si sa
 * composition porte UNE VIDÉO, UN AUDIO, ou AU MOINS DEUX IMAGES. Une image
 * seule, un document ou un lieu restent un post de base (POST).
 *
 * Directive durée minimale : une vidéo ou un audio ne qualifie QUE si sa
 * durée est CONNUE (en millisecondes) et >= `MIN_QUALIFYING_DURATION_MS`
 * (3s). Une durée absente/nulle est traitée comme non-qualifiante (jamais un
 * fallback permissif). Les images ne sont jamais soumises à cette condition.
 *
 * Miroir EXACT du SDK iOS : `ReelComposition.qualifiesAsReel`
 * (packages/MeeshySDK/Sources/MeeshySDK/Models/FeedModels.swift). Toute
 * évolution de la règle touche les deux sites — c'est la seule doctrine.
 *
 * Consommateurs : `PostService.createPost`/`updatePost` (gateway, import
 * direct), `PostComposer` (web, classification client avant publication) et
 * le backfill `scripts/migrations/reclassify-nonqualifying-reels-to-post.ts`.
 */

export const MIN_QUALIFYING_DURATION_MS = 3000;

export type ReelMediaLike = { mimeType: string | null; duration?: number | null };

function meetsMinDuration(duration: number | null | undefined): boolean {
  return typeof duration === 'number' && duration >= MIN_QUALIFYING_DURATION_MS;
}

export function qualifiesAsReel(media: ReadonlyArray<ReelMediaLike>): boolean {
  const normalized = media.map((m) => ({
    mime: (m.mimeType ?? '').toLowerCase(),
    duration: m.duration,
  }));
  const hasQualifyingVideo = normalized.some((m) => m.mime.startsWith('video/') && meetsMinDuration(m.duration));
  const hasQualifyingAudio = normalized.some((m) => m.mime.startsWith('audio/') && meetsMinDuration(m.duration));
  const imageCount = normalized.filter((m) => m.mime.startsWith('image/')).length;
  return hasQualifyingVideo || hasQualifyingAudio || imageCount >= 2;
}

File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/ReelComposition.kt
package me.meeshy.sdk.model

/**
 * Creation-time reel-qualification predicate — pure port of iOS SDK
 * `ReelComposition` (packages/MeeshySDK/Sources/MeeshySDK/Models/FeedModels.swift)
 * and mirror of the gateway's authoritative `qualifiesAsReel`
 * (services/gateway/src/services/posts/reelComposition.ts).
 *
 * Règle produit (directive user 2026-08-02) : un post n'est un RÉEL que si sa
 * composition porte une vidéo, un audio, ou au moins deux images. Une image
 * seule, un document ou un lieu restent un post de base (POST).
 *
 * Directive durée minimale : une vidéo ou un audio ne qualifie QUE si sa durée
 * est CONNUE et >= [MIN_QUALIFYING_DURATION_MS] (3s). Une durée absente/nulle
 * est traitée comme non-qualifiante — jamais un fallback permissif. Les
 * images ne sont jamais soumises à cette condition.
 *
 * The gateway is the final authority (`PostService.createPost` silently
 * degrades a claimed REEL that doesn't qualify back to POST) — this predicate
 * exists so the Android client requests the correct [PostType] up front
 * instead of always guessing POST, exactly like iOS's `FeedView.composerOverlay`
 * + `composerForcePlainPost` toggle. Any evolution of the rule touches all
 * three sites (iOS SDK, gateway, here) — that is the single doctrine.
 */
public object ReelComposition {
    public const val MIN_QUALIFYING_DURATION_MS: Long = 3000

    /** One item's composition-relevant shape: MIME type + known duration (video/audio only). */
    public data class MediaKind(val mimeType: String, val durationMs: Long? = null)

    /**
     * `true` when a post with this [media] composition qualifies as a reel: a
     * qualifying video or audio (duration known and >= [MIN_QUALIFYING_DURATION_MS]),
     * or at least two images. Documents and every other kind never qualify —
     * and neither does a single image or a video/audio whose duration is too
     * short or unknown.
     */
    public fun qualifiesAsReel(media: List<MediaKind>): Boolean {
        val normalized = media.map { it.mimeType.lowercase() to it.durationMs }
        val hasQualifyingVideo = normalized.any { (mime, duration) -> mime.startsWith("video/") && meetsMinDuration(duration) }
        val hasQualifyingAudio = normalized.any { (mime, duration) -> mime.startsWith("audio/") && meetsMinDuration(duration) }
        val imageCount = normalized.count { (mime, _) -> mime.startsWith("image/") }
        return hasQualifyingVideo || hasQualifyingAudio || imageCount >= 2
    }

    private fun meetsMinDuration(durationMs: Long?): Boolean =
        durationMs != null && durationMs >= MIN_QUALIFYING_DURATION_MS

    /**
     * The default [PostType] for a new post: [PostType.REEL] when [media]
     * qualifies and the author hasn't [forcePlainPost]d, otherwise [PostType.POST].
     */
    public fun defaultType(media: List<MediaKind>, forcePlainPost: Boolean = false): PostType =
        if (!forcePlainPost && qualifiesAsReel(media)) PostType.REEL else PostType.POST
}
